RecoTauTag/RecoTau/plugins/DeepTauId.cc

```diff
@@ -1963,215 +2054,245 @@ class DeepTauId : public deep_tau::DeepTauBase {
+      get(dnn::rho) = sp.scale(rho, dnn::rho);
+      get(dnn::tau_pt) = sp.scale(tau.polarP4().pt(), dnn::tau_pt);
+      get(dnn::tau_eta) = sp.scale(tau.polarP4().eta(), dnn::tau_eta);
+      get(dnn::tau_inside_ecal_crack) = sp.scale(isInEcalCrack(tau.polarP4().eta()), dnn::tau_inside_ecal_crack);
     }
     if (valid_index_pf_ele) {
+      const sc::ScalingParams& sp = scalingParamsMap_->at(std::make_pair(ft_PFe, is_inner));
       size_t index_pf_ele = cell_map.at(CellObjectType::PfCand_electron);
       const auto& ele_cand = dynamic_cast<const CandidateCastType&>(pfCands.at(index_pf_ele));
 
-      get(dnn::pfCand_ele_valid) = valid_index_pf_ele;
-      get(dnn::pfCand_ele_rel_pt) = getValueNorm(pfCands.at(index_pf_ele).polarP4().pt() / tau.polarP4().pt(),
-                                                 is_inner ? 0.9792f : 0.304f,
-                                                 is_inner ? 0.5383f : 1.845f);
-      get(dnn::pfCand_ele_deta) = getValueLinear(pfCands.at(index_pf_ele).polarP4().eta() - tau.polarP4().eta(),
-                                                 is_inner ? -0.1f : -0.5f,
-                                                 is_inner ? 0.1f : 0.5f,
-                                                 false);
-      get(dnn::pfCand_ele_dphi) = getValueLinear(dPhi(tau.polarP4(), pfCands.at(index_pf_ele).polarP4()),
-                                                 is_inner ? -0.1f : -0.5f,
-                                                 is_inner ? 0.1f : 0.5f,
-                                                 false);
-      get(dnn::pfCand_ele_pvAssociationQuality) =
-          getValueLinear<int>(candFunc::getPvAssocationQuality(ele_cand), 0, 7, true);
-      get(dnn::pfCand_ele_puppiWeight) = is_inner ? getValue(candFunc::getPuppiWeight(ele_cand, 0.9906834f))
-                                                  : getValue(candFunc::getPuppiWeight(ele_cand, 0.9669586f));
-      get(dnn::pfCand_ele_charge) = getValue(ele_cand.charge());
-      get(dnn::pfCand_ele_lostInnerHits) = getValue<int>(candFunc::getLostInnerHits(ele_cand, 0));
-      get(dnn::pfCand_ele_numberOfPixelHits) = getValueLinear(candFunc::getNumberOfPixelHits(ele_cand, 0), 0, 10, true);
-      get(dnn::pfCand_ele_vertex_dx) =
-          getValueNorm(pfCands.at(index_pf_ele).vertex().x() - pv.position().x(), 0.f, 0.1221f);
-      get(dnn::pfCand_ele_vertex_dy) =
-          getValueNorm(pfCands.at(index_pf_ele).vertex().y() - pv.position().y(), 0.f, 0.1226f);
-      get(dnn::pfCand_ele_vertex_dz) =
-          getValueNorm(pfCands.at(index_pf_ele).vertex().z() - pv.position().z(), 0.001f, 1.024f);
-      get(dnn::pfCand_ele_vertex_dx_tauFL) = getValueNorm(
+      get(dnn::pfCand_ele_valid) = sp.scale(valid_index_pf_ele, dnn::pfCand_ele_valid - PFe_index_offset);
+      get(dnn::pfCand_ele_rel_pt) = sp.scale(pfCands.at(index_pf_ele).polarP4().pt() / tau.polarP4().pt(),
```

Comment:
pfCands.at is repeating several times in various forms in this file. Each time it performs bounds checking and can possibly throw an exception, adding overhead.

Moreover, it's not clear to me if `pfCands.at(index_pf_ele).polarP4()` (and the like) is recomputing some expensive trig every time?

Perhaps you can cache some stuff here?

Reply:
Indeed the amount of `at` calls can be reduced.

`polarP4` return stored `PolarLorentzVector` for both [PackedCandidate](https://github.com/cms-sw/cmssw/blob/master/DataFormats/PatCandidates/interface/PackedCandidate.h#L444)  and [PFCandidate](https://github.com/cms-sw/cmssw/blob/master/DataFormats/Candidate/interface/LeafCandidate.h#L116). That is the exact reason why `polarP4` is used instead of `p4` - to avoid expensive computations at each call.

Reply:
It is correct that `pfCands.at(index)` used several time introduces some overhead, it can be easy improved by introducing a const ref to the pfcand at this index. But do we want to do it in this PR? I also would like to note that this part of code is not really new, it was just rewritten using more universal interface for scaling/standardization. 
What concerns `pfCand.polarP4()` - I think that it is cached internally within pat::PackedCandidate, i.e. p4 is built only once when the method is called first time by any module processing the Event.

Reply:
Reduction of number of `container.at(idx)` calls is reduced with [71bf7af](https://github.com/cms-sw/cmssw/pull/37892/commits/71bf7af2c86c299d87ffd6cf5be806973420cd72). In several places where it is safe by construction a direct access with `container[idx]` operator is used in other a "cache" (cost ref to often accessed element) is used.